src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderEnvironment.ValueTypeSupport.cs

```diff
@@ -0,0 +1,181 @@
+                    object thisField = RuntimeAugments.RhBoxAny(ref Unsafe.Add(ref thisRawData, fieldOffset), fieldType);
+                    object thatField = RuntimeAugments.RhBoxAny(ref Unsafe.Add(ref thatRawData, fieldOffset), fieldType);
+
+                    if (thisField == null)
+                    {
+                        if (thatField != null)
+                            return false;
+                    }
+                    else if (!thisField.Equals(thatField))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            //
+            // If the field layout information isn't present, we can memcompare
+            //
+
+            // Sanity check
+            Debug.Assert(RuntimeAugments.GetGCDescSize(typeHandle) == 0);
+
+            int valueTypeSize = typeHandle.GetValueTypeSize();
+            for (int i = 0; i < valueTypeSize; i++)
+                if (Unsafe.Add(ref thisRawData, i) != Unsafe.Add(ref thatRawData, i))
+                    return false;
+
+            return true;
+        }
+
+        public int ValueTypeGetHashCode(ValueType thisObj)
+        {
+            RuntimeTypeHandle typeHandle = RuntimeAugments.GetRuntimeTypeHandleFromObjectReference(thisObj);
+
+            if (!RuntimeAugments.IsDynamicType(typeHandle))
+            {
+                return StaticValueTypeGetHashCode(typeHandle, thisObj);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private int StaticValueTypeGetHashCode(RuntimeTypeHandle typeHandle, ValueType thisObj)
+        {
+            // TODO: we should also use hashcode of the type
+
+            NativeFormatModuleInfo module = ModuleList.GetModuleInfoByHandle(RuntimeAugments.GetModuleFromTypeHandle(typeHandle));
+
+            NativeHashtable layoutHashTable;
+            ExternalReferencesTable externalReferencesLookup;
+
+            if (!GetHashtableFromBlob(module, ReflectionMapBlob.InstanceFieldLayoutHashtable, out layoutHashTable, out externalReferencesLookup))
+                Environment.FailFast("Instance field layout not generated");
+
+            ref byte thisRawData = ref RuntimeAugments.GetRawData(thisObj);
+
+            var enumerator = layoutHashTable.Lookup(typeHandle.GetHashCode());
+
+            //
+            // First try to find instance field layout information in the hashtable.
+            //
+            NativeParser entryParser;
+            while (!(entryParser = enumerator.GetNext()).IsNull)
+            {
+                RuntimeTypeHandle tentativeType = externalReferencesLookup.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                if (!typeHandle.Equals(tentativeType))
+                    continue;
+
+                // Found the entry - find the first non-null field
```

Comment:
I believe this is only supposed to return the hashcode of the first field, whether or not it's null (yes, it's not a good hash code).

Reply:
> I believe this is only supposed to return the hashcode of the first field, whether or not it's null

That's not how CoreCLR does it:

https://github.com/dotnet/coreclr/blob/4e6435be1f5a77abb875f946fa7bd5dbeb5c3864/src/vm/comutilnative.cpp#L2158-L2159

Reply:
On closer inspection, you're right on both counts.

Reply:
If we have to match CoreCLR's behavior, we would also need to incorporate a type differentiator that gets XORed in. While eeTypePtr would be an easy one, I'd have concerns about leaking pointers. The CoreCLR code is:
```
    // To get less colliding and more evenly distributed hash codes,
    // we munge the class index with two big prime numbers
    hashCode = typeID * 711650207 + 2506965631U;
```

Reply:
Yup, that's why I have a TODO 24 lines above. I would probably use the type's stable hashcode that we include as a field of each EEType but I didn't think about how to combine it yet:

https://github.com/MichalStrehovsky/corert/blob/eee4c66cec59f3e93aa2e476b8e0cae1fa5af223/src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderEnvironment.ValueTypeSupport.cs#L120